src/autoreplace_cmd.cpp

```diff
@@ -470,6 +471,7 @@ static CommandCost ReplaceFreeUnit(Vehicle **single_unit, DoCommandFlag flags, b
 			*single_unit = new_v;
 
 			AI::NewEvent(old_v->owner, new ScriptEventVehicleAutoReplaced(old_v->index, new_v->index));
+			Game::NewEvent(new ScriptEventVehicleAutoReplaced(old_v->index, new_v->index));
```

Comment:
I'd almost template creating a new event in a manner that creates two events to keep the practical code duplication relatively low.

That would be something like:
```
class Script {
public:
    template <class EventClass, typename ... Args> 
    static void NewEvent(int owner, Args... args) {
        AI::NewEvent(owner, new EventClass(args...));
        Game::NewEvent(new EventClass(args...));
    }
};
```
and then `Script::NewEvent<ScriptEventVehicleAutoReplaced>(old_v->owner, old_v->index, new_v->index);`. Yes, when a company needs to be passed to the constructor you pass the owner twice in a row, or you make a second templated function `Script::NewEventWithOwner` or something that under water passes the owner twice to `Script::NewEvent`.

I'm not sure whether either way with the templating is the best solution though.

Reply:
I created a new commit at the end, adding 2 templates. I didn't know where to put this, so I created a new file just for that. Not sure if the file name is fine, or if this could be put in some existing file. Please let me know.

I have however, been unable to do it properly for the case of AI::BroadcastNewEvent, which has a default parameter.
https://github.com/OpenTTD/OpenTTD/blob/59251d3c6b9bf2a0ae85aefc2c9e9f91a46f3031/src/ai/ai.hpp#L117-L120

```
	/**
	 * Broadcast a new event to all active AIs, and to the Game Script.
	 * @param skip_company The company to skip broadcasting for, or MAX_COMPANIES to skip none.
	 */
	template <class ScriptEventType, typename ... Args>
	static void BroadcastNewEvent(CompanyID skip_company, Args ... args) {
		AI::BroadcastNewEvent(new ScriptEventType(args...), skip_company);
		Game::NewEvent(new ScriptEventType(args...));
	}
```

When there is a `skip_company`, such in the case of a new company, I pass `Script::BroadcastNewEvent<ScriptEventCompanyNew>(c->index, c->index);`, but in the case I want to broadcast to all companies, I need to pass `MAX_COMPANIES`, which is the majority of the situations. One such example is `Script::BroadcastNewEvent<ScriptEventCompanyInTrouble>(MAX_COMPANIES, c->index);`
Ideally I wouldn't have to use `MAX_COMPANIES` everywhere, but I don't know how to do it in the template.